src/VisualStudio/Core/Def/Workspace/VisualStudioSourceGeneratorTelemetryCollectorWorkspaceServiceFactory.cs

```diff
@@ -0,0 +1,105 @@
+using Microsoft.CodeAnalysis.Host;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.Internal.Log;
+using Microsoft.CodeAnalysis.SourceGeneratorTelemetry;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Shell.Interop;
+
+namespace Microsoft.VisualStudio.LanguageServices
+{
+    /// <summary>
+    /// Exports a <see cref="ISourceGeneratorTelemetryCollectorWorkspaceService"/> which is watched across all workspaces. This lets us collect
+    /// statistics for all workspaces (including things like interactive, preview, etc.) so we can get the overall counts to report.
+    /// </summary>
+    [Export]
+    [ExportWorkspaceServiceFactory(typeof(ISourceGeneratorTelemetryCollectorWorkspaceService)), Shared]
+    internal class VisualStudioSourceGeneratorTelemetryCollectorWorkspaceServiceFactory : IWorkspaceServiceFactory, IVsSolutionEvents
+    {
+        /// <summary>
+        /// The collector that's used to collect all the telemetry for operations within <see cref="VisualStudioWorkspace"/>. We'll report this
+        /// when the solution is closed, so the telemetry is linked to that.
+        /// </summary>
+        private readonly SourceGeneratorTelemetryCollectorWorkspaceService _visualStudioWorkspaceInstance = new SourceGeneratorTelemetryCollectorWorkspaceService();
+
+        /// <summary>
+        /// The collector used to collect telemetry for any other workspaces that might be created; we'll report this at the end of the session since nothing here is necessarily
+        /// linked to a specific solution. The expectation is this may be empty for many/most sessions, but we don't want a hole in our reporting and discover that the hard way.
+        /// </summary>
+        private readonly SourceGeneratorTelemetryCollectorWorkspaceService _otherWorkspacesInstance = new SourceGeneratorTelemetryCollectorWorkspaceService();
+
+        private readonly IThreadingContext _threadingContext;
+        private readonly IAsyncServiceProvider _serviceProvider;
+        private volatile int _subscribedToSolutionEvents;
+
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public VisualStudioSourceGeneratorTelemetryCollectorWorkspaceServiceFactory(IThreadingContext threadingContext, SVsServiceProvider serviceProvider)
+        {
+            _threadingContext = threadingContext;
+            _serviceProvider = (IAsyncServiceProvider)serviceProvider;
+        }
+
+        public IWorkspaceService CreateService(HostWorkspaceServices workspaceServices)
+        {
+            // We will record all generators for the main workspace in one bucket, and any other generators running in other
+            // workspaces (interactive, for example) will be put in a different bucket. This allows us to report the telemetry
+            // from the primary workspace on solution closed, while not letting the unrelated runs pollute those numbers.
```

Comment:
note: i think we are not supposed to do solution-close anymore because it slows down VS close.  we've had to roll back several bits of reporting because of that.

Reply:
(resolving in favor of other conversation)